### ISSUE TITLE
@backstage/plugin-cost-insights 0.11.13 fails to load correctly in tests

### ISSUE THREAD

**Issue by JP-Dhabolt** [backstage/backstage#8554]
## Expected Behavior

Upgrading `@backstage/plugin-cost-insights` should not cause any issues.

## Current Behavior

Upgrading `@backstage/plugin-cost-insights` from `0.11.12` to `0.11.13` causes test suite failure, with the error:
```
TypeError: Cannot read property 'default' of undefined

      1 | import * as luxon from 'luxon';
    > 2 | import { Duration } from '@backstage/plugin-cost-insights';
        |                           ^
      3 | import { DEFAULT_COST_DATE_FORMAT } from './format';
      4 | import { assertNever } from './assert';
      5 |

      at ../../../node_modules/node_modules/react-dom/cjs/react-dom.development.js:18:13
      at Object.<anonymous> (../../../node_modules/node_modules/react-dom/cjs/react-dom.development.js:25011:4)
      at Object.<anonymous> (../../../node_modules/@backstage/plugin-cost-insights/dist/index.esm.js:1:348)
      at Object.<anonymous> (utils/duration.ts:2:27)
      at Object.<anonymous> (clients/costInsights/dummyUtils.ts:23:17)
      at Object.<anonymous> (clients/costInsights/costInsightsClient.ts:14:19)
      at Object.<anonymous> (clients/costInsights/costInsightsClient.test.ts:3:27)
```

While the tests fail, the build itself succeeds, and the application can run without any errors.

## Possible Solution

It looks like in commit https://github.com/backstage/backstage/commit/be32bc3edbe758dd864f829142023904ba65944a, `react-dom` was removed from the `dependencies` in `plugins/cost-insights/package.json`.

Looking at the dependency tree, there are multiple sub-dependencies of this plugin that require `react-dom` as a peer.  I am by no means an expert on peer dependencies, but it is my understanding that not re-declaring transitive peer dependencies can sometimes cause issues.

My best guess is that there are some resolution issues happening when using `ts-node` (via the backstage-cli), since the issue only occurs during testing, and not while the application is actually running (when `ts-node` is no longer in the picture).

One other thing I noticed, though I don't think it is the cause of this issue, is that some of the downstream dependencies for the `@backstage/plugin-cost-insights` plugin have peerDependencies requirements for React v15/v16, meaning that if v17 were to actually be used, it would potentially have errors, as it does not match downstream peer dependencies.

## Steps to Reproduce

1. Update `@backstage/plugin-cost-insights` from `0.11.12` to `0.11.13`
2. Run test on files that import anything from the plugin (3 separate test files are failing, all with the same error, importing different parts of the plugin).

## Context

* Originally trying to do a version bump with `backstage-cli versions:bump`.  Narrowed the problem down to just the one plugin.
* When the plugin is updated, no other entry in the `yarn.lock` is changed, as the new plugin dependencies are already present from other plugins.
* `react-dom` is installed in the app, and `yarn why` shows only one version is installed (`react-dom@16.14.0`).

## Your Environment

- NodeJS Version (v14): `v14.17.6`
- Operating System and Version (e.g. Ubuntu 14.04): `Linux 5.4.0-1063-azure - linux/x64`
- Browser Information: N/A

**Comment by kuangp** [comment]
I also ran into this and have been banging my head against the wall for a while. Is this any current workarounds to get our tests passing?

**Comment by freben** [comment]
Hmm. The error you pasted was from Duration, which is an `enum` - i hear those are finicky and can lead to problems. @timboniT I'm cloud be worth looking into making that just an object or a set of `'X' | 'Y' |...` strings

**Comment by kuangp** [comment]
wanted to point out that as @JP-Dhabolt noted, the `Cannot read property 'default' of undefined` error shows up for anything you import from the cost insights plugin and is not just specific to `Duration`

**Comment by Rugvip** [comment]
Wondering if `react` is also installed, also with a single version, `16.14.0`? Afaik React can be pretty picky about having those two sync'd up.

The path in your stack trace looks a bit weird with the double `node_modules`, does it look like that in your file structure as well? that is `../../../node_modules/node_modules/react-dom/cjs/react-dom.development.js:18:13`

**Comment by JP-Dhabolt** [comment]
@Rugvip I'm OOO until after the New Year, so I can't verify the `react` version until next week.  I do know we only have 1 version installed where we have control (`16.x`), though IIRC there are some deeper dependencies (from Backstage plugins/packages) that rely on `17.x` that are unrelated to this module.

To answer the question on the whether the file structure looks like that, no it does not.  I believe this is why it cannot find `react-dom`, hence the import error.  This is what led me down the road of looking at the removal of `react-dom` as an explicit dependency (it seems too much to be a coincidence that the error started happening once the explicit dependency was removed, and the explicit peer dependency on `react-dom` was not added).

The double `node_modules` seems like it could be a dependency hoisting issue, but again, one that is only occurring within the tests.  When actually building/running Backstage, there are no import errors, only when running the tests via the Backstage CLI.  If I was following the code for the CLI commands correctly, this uses `ts-node` under the hood, meaning whatever differences there are in dependency resolution from `ts-node` vs whatever build tools are used is where this issue is raising its head.

**Comment by Rugvip** [comment]
@JP-Dhabolt tests are run by jest, so it should be [this](https://github.com/backstage/backstage/blob/8e5d716c0acf7e2c6e873d23af18b9286daeea2c/packages/cli/config/jest.js#L104) configuration that is handling things rather than `ts-node`. I'll give reproducing this a go in a new app and see where I end up.

**Comment by Rugvip** [comment]
Able to reproduce and done some digging, and it seems like this is a build system issue. The published cost-insights plugin seemingly has all its dependencies bundled into the dist script, which will ofc break because there's for example a duplicate react and a bunch of other fun stuff in there.

**Comment by Rugvip** [comment]
TL;DR: all of `@backstage/test-utils` was being built into the published package. Bit more details in the fix: #8644

**Comment by Rugvip** [comment]
Tbh I think the best workaround for now is likely to leave cost-insights at `0.11.12`, don't really want to make a patch release for this one

**Comment by JP-Dhabolt** [comment]
Thanks @Rugvip . I'm good with the workaround. Any chance of unpublishing the impacted version so running backstage bump doesn't keep trying to pick it up?

**Comment by Rugvip** [comment]
I'm afraid it's a bit too late to unpublish at this point, but I added deprecation messages for the two broken releases
